piet-direct2d/src/dwrite.rs

```diff
@@ -133,42 +151,114 @@ impl DwriteFactory {
+            let mut exists = 0_i32;
+            let mut locale_name = [0_u16; LOCALE_NAME_MAX_LENGTH];
+
+            let success =
+                GetUserDefaultLocaleName(locale_name.as_mut_ptr(), LOCALE_NAME_MAX_LENGTH as i32);
+            let mut hr = if SUCCEEDED(success) {
+                names.FindLocaleName(locale_name.as_ptr(), &mut index, &mut exists)
+            } else {
+                // we reuse the previous success; we want  to run the next block
+                // both if the previous failed, or if it just doesn't find anything
+                hr
+            };
+            if !SUCCEEDED(hr) || exists == 0 {
+                hr = names.FindLocaleName(DEFAULT_LOCALE.as_ptr(), &mut index, &mut exists);
+            }
+
+            if !SUCCEEDED(hr) {
+                return Err(hr.into());
+            }
+
+            // if locale doesn't exist, just choose the first
+            if exists == 0 {
+                index = 0;
+            }
+
+            let mut length = 0_u32;
+            let hr = names.GetStringLength(index, &mut length);
 
-    pub fn family(mut self, family: String) -> TextFormatBuilder {
-        self.family = Some(family);
-        self
+            if !SUCCEEDED(hr) {
+                return Err(hr.into());
+            }
+
+            let mut wide_name: Vec<u16> = Vec::with_capacity(length as usize + 1);
+            let hr = names.GetString(index, wide_name.as_mut_ptr(), length + 1);
+            if SUCCEEDED(hr) {
+                wide_name.set_len(length as usize + 1);
+                let name = OsString::from_wide(&wide_name)
+                    .into_string()
+                    .unwrap_or_else(|err| err.to_string_lossy().into_owned())
+                    .into();
+                let wide_name = Arc::from(wide_name);
+
+                Ok(FamilyName { name, wide_name })
+            } else {
+                Err(hr.into())
+            }
+        }
     }
+}
+
+impl TextFormat {
+    pub(crate) fn new(
+        factory: &DwriteFactory,
+        family: impl AsRef<[u16]>,
+        size: f32,
+    ) -> Result<TextFormat, Error> {
+        let family = family.as_ref();
 
-    pub fn build(self) -> Result<TextFormat, Error> {
-        let family = self
-            .family
-            .expect("`family` must be specified")
-            .to_wide_null();
-        let size = self.size.expect("`size` must be specified");
-        let locale = "en-US".to_wide_null();
         unsafe {
             let mut ptr = null_mut();
-            let hr = self.factory.0.CreateTextFormat(
+            let hr = factory.0.CreateTextFormat(
                 family.as_ptr(),
                 null_mut(), // collection
                 DWRITE_FONT_WEIGHT_NORMAL,
                 DWRITE_FONT_STYLE_NORMAL,
                 DWRITE_FONT_STRETCH_NORMAL,
                 size,
-                locale.as_ptr(),
+                //TODO: this should be the user's locale? It will influence font fallback behaviour?
```

Comment:
Yes, this should be the user's locale. This is a fairly complex topic, and being able to override the locale on a per-span basis should be a feature at some point, but this is ok for now. The main thing that will be impacted is CJK.

Reply:
A lot of plumbing there to figure out, since we'll want the locale to be part of the general piet API in this case, likely as an attribute on the layout; going to punt to a future PR.